### PR TITLE
Add mackerel-plugin-gunicorn

### PR DESCRIPTION
**Pull request by kizkoh**
This plugin is inspired by mackerel-plugin-plack.

Many monitoring SaaS tool support gunicorn monitoring using statsd daemon.  But, Gunicorn send the number of all worker to statd server.  This plugin support to distinguish active worker from idle worker without statsd daemon.

### CODE REVIEW

File: mackerel-plugin-gunicorn/README.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+## Synopsis
+
+```shell
+mackerel-plugin-gunicorn [-status-file=<path>]
+```
+
+## Requirements
+
+This plugin require that gunicorn include the following code in its `config.py`.
```
Comment (itchyny) [inline mackerel-plugin-gunicorn/README.md]: [nit] require -> requires

File: mackerel-plugin-gunicorn/lib/gunicorn.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+//       "status": "_"
+//     },
+//     {
+//       "host": "localhost:8000",
+//       "method": "GET",
+//       "uri": "/",
+//       "pid": 26724,
+//       "status": "_"
+//     }
+//   ],
```
Comment (itchyny) [inline mackerel-plugin-gunicorn/lib/gunicorn.go]: [nit] I think this is a JSON sample, remove trailing semicolon.

File: mackerel-plugin-gunicorn/README.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+SERVER_STATUS = "/dev/shm/gunicorn_status.json"
+
+
+def on_starting(servers):
+    if os.path.exists(SERVER_STATUS):
+        os.remove(SERVER_STATUS)
+    with open(SERVER_STATUS, mode="w") as f:
+        obj = {
+            "TotalAccesses": "0",
```
Comment (itchyny) [inline mackerel-plugin-gunicorn/README.md]: I'm curious why you don't use numbers for these three fields.

File: mackerel-plugin-gunicorn/README.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                for i, _ in stats:
+                    obj["stats"][i] = stat
+
+            obj["IdleWorkers"] = str(int(obj["IdleWorkers"]) + 1)
+            obj["BusyWorkers"] = str(int(obj["BusyWorkers"]) - 1)
+
+            f.seek(0)
+            f.truncate(0)
+            json.dump(obj, f)
+            f.flush()
```
Comment (itchyny) [inline mackerel-plugin-gunicorn/README.md]: Reading data from file, update and write. Isn't there any race condition?

Comment (itchyny) [inline mackerel-plugin-gunicorn/README.md]: Oops, flock seems to be the answer for that, sorry.

File: mackerel-plugin-gunicorn/README.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        fcntl.flock(f.fileno(), fcntl.LOCK_EX)
+        try:
+            obj = json.load(f)
+
+            stats = [(i, v) for i, v in enumerate(obj["stats"]) if v["pid"] == worker.pid]
+            if len(stats) == 0:
+                worker.log.warn("not find worker.pid: %d in stats object", worker.pid)
+            else:
+                for i, _ in stats:
+                    obj["stats"][i] = stat
```
Comment (itchyny) [inline mackerel-plugin-gunicorn/README.md]: When `obj["stats"]` has length 5 and `stats` has length 4, `obj["stats"][4]` will not be updated nor be deleted. Is this an intended behaviour?

Comment (itchyny) [inline mackerel-plugin-gunicorn/README.md]: Oops, I was wrong... Maybe this codes intends to update the stats with the same pid, sorry.

Comment (kizkoh) [inline mackerel-plugin-gunicorn/README.md]: Yes, it is an element that is equal to pid of worker in obj["stats"] to want to update it.  It is `stat` and its length is 1.

Comment (itchyny) [inline mackerel-plugin-gunicorn/README.md]: When you want to update the single stat with the same pid, I think it's better if the code looks like... `index = find_index(...); if index < 0 then {append} else {assign}`. I think `for i, _ in stats` looks as if there might be multiple occurrence of stat with the same pid.

**Comment by kizkoh** [comment]
This plugin requires to write complex python code to gunicorn config.
This problem should be settled in WSGI middleware. I close this pull request.

**Pull request closed without merging**